### PR TITLE
Towards AutoMOC - Added: File name cleaning, auto-tagging, new settings - Removed: Index File/Folder no longer listed

### PR DESCRIPTION
**Pull request by MichaBrugger**
Hey akos,

We've briefly spoken via email, so here I am 😉 
This is my first-ever pull request, so I might have overdone it a bit with the documentation in the README, but I hope you can forgive me, I just thought I'd rather write it all out.

Anyway, lets get to it:
- Generally, the changes I implemented go away a bit from the thought of pure "indexing" and maybe more towards the MOC-thought. Basically steering less towards seeing what's inside a folder and more towards the idea that things are automatically mapped and remapped according to their broad topic (hence the name AutoMOC). I'm obviously not sure if this is the vision that you had for the plugin. But since you laid the full "groundwork" (you basically did all the difficult parts with the file-creation etc.) for it I didn't want to take anything "away" from you and I would love to combine these two ideas into an even better plugin that would surely help many people 🙌🏼

- So, long story short it's really up to you, if you like the idea then let's combine ideas (I have to say I like the name AutoMOC a bit better, but that would be up for discussion 🤣). And if you want to go in another direction, that's fine too. Just let me know and I'll keep working on it by myself.

- Regarding the actual changes: I documented the whole idea and functionality in the README and you can obviously see the changes in the code. So there is not much I can add here.

I hope you like it and I'm looking forward to hearing from you!
Best regards from Switzerland,

Micha

### CODE REVIEW

**Comment by akosbalasko** [comment]
Huhh, @MichaBrugger I'm so sorry, I totally missed your PR notification. I'll check your changes asap (probably tomorrow)!

**Comment by MichaBrugger** [comment]
@akosbalasko, no issue at all! To be honest, I'm not that sure anymore if the ideas I had were as good as I thought at the time haha

But anyway, I'll let you be the judge of that! Looking forward to your feedback :)

File: README.md [hunk 1]
```diff
@@ -1,65 +1,50 @@
+# Obsidian AutoMOC
+#### Based on [Zoottelkeeper](https://github.com/akosbalasko/zoottelkeeper-obsidian-plugin) 0.6.0
 
+**This currently works seperate from Zoottelkeeper but I'm hoping to integrate it with the main plugin.**
 
-# ZoottelKeeper: A Zookeeper for your Zettelkasten folder
```
Comment (akosbalasko) [inline README.md:3]: Don't you like the name? :D Your's is definitely more self-explainable (?) but I think we cannot change the name of the package and the plugin just like that, because we loose the continuity with the previous versions in Obsidian. May I ask you to switch it back? Of course, with keeping you among the authors, as I tested you did an amazing job! :)

File: main.ts [hunk 2]
```diff
@@ -1,234 +1,334 @@
...
-		return (fileFolderArray[0] !== '') ? fileFolderArray[fileFolderArray.length - 1] : this.app.vault.getName();
-	}
-
-	isIndexFile = (file: TAbstractFile): boolean => {
-
-		return file instanceof TFile && file.name.startsWith(this.settings.indexPrefix);
-
-	}
-
+  settings: ZoottelkeeperPluginSettings;
```
Comment (akosbalasko) [inline main.ts:30]: I think the lint config is different , because the tabs are replaced by double-spaces, and therefore it's really hard to see what are the real changes. Could you please change the double-spaces indentations to tabs? Thank you!

File: manifest.json [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
-	"name": "Zoottelkeeper Plugin",
-	"version": "0.6.0",
+	"id": "obsidian_autoMOC",
+	"name": "AutoMOC",
+	"version": "0.0.2",
 	"minAppVersion": "0.12.1",
-	"description": "Zoottelkeeper creates and maintains index files in all of your folders in your vault. If you create/delete/move a note, the index files will be updated automatically.",
-	"author": "Akos Balasko",
-	"authorUrl": "https://github.com/akosbalasko",
+	"description": "This plugin automatically creates, maintains and tags MOCs for all your folders. It is based on Zoottelkeeper 0.6.0.",
```
Comment (akosbalasko) [inline manifest.json:6]: We can keep this, without the last sentence of course :)

File: manifest.json [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
+	"author": "Akos Balasko, Micha Brugger",
```
Comment (akosbalasko) [inline manifest.json:7]: welcome onboard! :)

File: package.json [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
-  "author": "Akos Balasko",
+  "keywords": [
+    "zettelkasten",
+    "obsidian.md",
+    "obsidian-plugin"
+  ],
+  "author": "Akos Balasko, Micha Brugger",
   "license": "MIT",
   "devDependencies": {
     "@rollup/plugin-commonjs": "^18.0.0",
```
Comment (akosbalasko) [inline package.json:18]: could you please remove the hats(^) from the dependency versions? If we specify precisely the required versions of the dependencies, we prevent us from break our stuff by some bugs may be added in the upcoming versions there.

**Review by akosbalasko — CHANGES_REQUESTED** [review]
In general: it looks super cool, and of course, I would like to integrate your amazing job into the main plugin. 

Some issues, that I realized during the testing, and they look a bit serious:

1. If I rename a folder, its inner indexFile won't be renamed (and the graph breaks) 
2. File/Folder creations not, just file rename trigger regeneration of the index files
3. It looks like that the regeneration of the index files breaks the chain if I move a complete subfolder into another subfolder if both have the same names. 

I'd investigate it a bit deeper, but first of all, please set your indentation to tabs to let me see the real changes in this PR.
It looks like that many of the problems are related to the improvement of switching relative paths from absolute ones, but it's still just a tip. 

Thank you for all your hard work, it will be a great improvement! at least 3 mid. semver version upgrade, so 0.9.0, we're coming! :)

File: README.md (hunk 1 above)
Comment (MichaBrugger) [reply README.md:3]: Yes! I'll switch it back! To be honest, I changed it because I wasn't sure you'd accept the PR. Since you like the changes, I think we should really just stick with Zoottelkeeper and I'll let everything there the way it was 👍🏼

File: main.ts (hunk 2 above)
Comment (MichaBrugger) [reply main.ts:30]: Oh! Yes, I'll change that. Should I make a new PR afterwards? (Sorry if that's a stupid question, it's my first 😅)

File: manifest.json (hunk 4 above)
Comment (MichaBrugger) [reply manifest.json:7]: Haha, I hope you don't take this the wrong way! It was done with the intent of not taking away any credit from you given that it might stay a forked version. Now I feel like I'm taking way too much credit since you did the bulkload of the work.. We can also remove it for now.

File: package.json (hunk 5 above)
Comment (MichaBrugger) [reply package.json:18]: I will do that! Once again, I'm new to this and never did any "professional" programming, so I really appreciate when you just let me know whenever you notice things like that, I'll try to keep them in mind for next time 👍🏼

**Comment by MichaBrugger** [comment]
Generally: I will be working on solving the issues you mentioned over the weekend and will then make a new PR 🙌🏼 

Also, you mentioned a few times how the "creation" of new index-files doesn't happen in certain cases (moving a complete folder into a subfolder, etc). I noticed that too and I realized that the reason for it is, that the whole function is only triggered on "creation", "delete" and "rename", which makes it obvious, why a creation of a subfolder (or similar) wouldn't trigger the regeneration of the index file. However, so far I haven't found any possibility to add "moving" as a triggerevent. Do you have any idea how we could solve this?

File: main.ts (hunk 2 above)
Comment (akosbalasko) [reply main.ts:30]: Nope, this PR should be updated automatically if you push your changes to the forked master. There are no stupid questions btw :)

File: manifest.json (hunk 4 above)
Comment (akosbalasko) [reply manifest.json:7]: No no, you can keep yourself in, there are so many improvements in your PR, so I'm glad to share the credits with you!

File: package.json (hunk 5 above)
Comment (akosbalasko) [reply package.json:18]: No worries :)

**Comment by akosbalasko** [comment]
On your question about 'move' as a trigger event: as far as I remember that is solved in Obsidian by two events, a "delete" and a "create" right after each other, that was the main reason why I had to wait a while and do the modifications in a batch (done by `debounce` function in line 33, it runs the function in each 1000 ms, just once even if there were more trigger events pushed on that timeframe. 

Thank you very much, and good luck!

**Comment by MichaBrugger** [comment]
I've been looking into the whole thing with the broken chain: The reason is that while in higher level file the indexed file does not have the ".md" ending while in the lower level file it does have a ".md" ending (even though it's invisible). That's the reason it doesn't properly link up.

Example (these two files would link):
image

I'll try to solve that so it works with Clean File Path too 👍🏼

**Comment by akosbalasko** [comment]
Perfect, thank you!:

**Comment by MichaBrugger** [comment]
I'm struggling to solve the chain-breaking issue without creating new bugs. I'm giving my best to solve it whenever I have time, but I might take a few more days. If I don't manage to solve it by Friday, I would suggest doing a partial implementation of everything that now works (which is basically everything except the "Clean String option"). What do you think?

**Comment by akosbalasko** [comment]
Hi @MichaBrugger !

Sure sure, I agree, partial improvements are welcomed as well, they are improvements, and it makes no sense for me to block them because of a parallel, but more complicated improvement.

Thank you for your efforts!

**Comment by MichaBrugger** [comment]
I was able, as far as I'm able to tell, to solve the challenges we discussed, so I made a new commit. I hope you'll like it 🙌🏼 

Regarding indentation: I did change everything to tabs. But unfortunately, I have auto-formatting enabled and GitHub often labels things as "new" even though they are only arranged more nicely. Sorry about that :(

**Comment by akosbalasko** [comment]
Hi @MichaBrugger !
Thank you, I'll check it tomorrow.

**Comment by akosbalasko** [comment]
Hi @MichaBrugger !

Thank you very much for your great effort, I tested your changes, and it looks like that we still have the following issues: 

1. renaming a complete (sub)folder does not rename the index file itself
2. If a folder contains a subfolder with the same name, then it still breaks the chain (I think it stands for the case if I name the index file same as the folder)
3. the chain still breaks if I have a file with the same name in two subfolders and in the parent folder as well. (I think its because of the relative paths)

Screenshot 2021-08-07 at 8 15 32
Screenshot 2021-08-07 at 8 16 40

I would suggest to split this PR to fine stuff and problematic stuff, I mean I gladly merge a PR that contains the autotagger feature and the cleanFiles as two different PRs. And the exclusion of the folder and the index file from the list. Could you please split these features and create a PR one per each? Thank you very much! 
The problematic is the relative folder linking and the index file naming after its parent folder, let's discuss it further on its specific PR .

Thanks a lot!

**Comment by MichaBrugger** [comment]
Hey Akos! First of all, thanks for testing and your feedback! I'm gladly going to split the PR (I'll do so today), but a few things I don't understand with the points you mentioned:

1. You're right, it doesn't rename the index-file, but it creates a new index file with the new folder name (I left it like that on purpose, because else I would have to delete the other file (?) and I thought this to be dangerous territory. Also, as far as I can tell in the current version of Zoottelkeeper this would just break the chain if I do it on a lower level folder.
image

2. Here I'm really confused because it's an error I can't reproduce. When I have files/nested subfolders that are named the same it looks like this (even with Clean Paths on):
image

3.  You're right here, that's a case that I didn't test for. It occurs if the there is a file with the same name in the "root" folder. And you're also right about the fact that this definitely is related to the clean path feature, so I'll go over that again and remove it from the current PR.
image
image

It might help if we could quickly chat at some point. But besides that, I'll give my best to solve these issues asap, thanks for taking your time for such a detailled feedback :)

File: README.md [hunk 6]
```diff
@@ -1,65 +1,62 @@
...
-
-Assume that we have a Zettelkasten folder with 2 subfolders `FolderA` and `FolderB`.
-
-1. If I create a note called `fileA.md` in `FolderA` then ZoottelKeeper creates an index file within `FolderA` called **_Index_of_FolderA.md** with content: 
-    - **[[fileA]]**
-    - **[[_Index_of_FolderA.md]]**
-
-2. If the file is being moved from `FolderA` to `FolderB`, **_Index_of_FolderB.md** is going to be created in `FolderB` with content
+### 2.1 Introduction
+So, the idea behind AutoMOC (and Zoottelkeeper) is to help you generate the base form of these maps automatically. It does so by indexing all the files and folders that lay in a folder, thus creating a link from the file to all it's content.
```
Comment (akosbalasko) [inline README.md:15]: Please improve the text by removing AutoMOC, because we should keep the original name.

File: README.md [hunk 7]
```diff
@@ -1,65 +1,62 @@
...
+You can choose to add YAML Meta Tags to your automatically generated index-files.
 
-As the folder links don't point to real notes (because they are folders), to make them visible in the Graph view, please enable the option of viewing orphan notes.
+### 3.5 Set Custom Meta Tags
+You can set one or multiple custom Meta Tags. Since they are displayed in the YAML format, you don't need to add a '#'.
+If you're setting multiple tags please make sure to separate them with commas.
 
+### 3.6 Additional Things
+- The indexed files are sorted alphabetically.
+- The file and the folder are no longer listed in the in the index-file. (This currently only works if you toggle 'Clean Files'.)
```
Comment (akosbalasko) [inline README.md:58]: I think with a solution in the other PR this feat. is going to be supported by default.

File: main.ts [hunk 8]
```diff
@@ -1,87 +1,118 @@
...
 	}
 	loadVault() {
-		this.lastVault = new Set(this.app.vault.getMarkdownFiles().map(file => file.path));
+		this.lastVault = new Set(
+			this.app.vault.getMarkdownFiles().map((file) => file.path)
+		);
 	}
 	async keepTheZooClean() {
-		console.debug('keeping the zoo clean...');
+		console.debug("keeping the zoo clean...");
```
Comment (akosbalasko) [inline main.ts:65]: pls keep the single quotes instead of apostrophes.

File: main.ts [hunk 9]
```diff
@@ -1,87 +1,118 @@
...
-
+		this.lastVault = new Set(
+			this.app.vault.getMarkdownFiles().map((file) => file.path)
+		);
 	}
 
 	onunload() {
-
-		console.debug('unloading plugin');
+		console.debug("unloading plugin");
```
Comment (akosbalasko) [inline main.ts:115]: here as well

File: main.ts [hunk 10]
```diff
@@ -94,99 +125,100 @@ export default class ZoottelkeeperPlugin extends Plugin {
...
 			indexContent.push(this.generateIndexItem(parentLink));
 		}
 		try {
-			if (indexTFile instanceof TFile)
-				await this.app.vault.modify(indexTFile, indexContent.join('\n'));
-			else {
-				throw new Error('Creation index as folder is not supported');
+			if (indexTFile instanceof TFile){
+				if (this.settings.indexTagBoolean === true) {
+					const tag = this.settings.indexTagValue.valueOf();
```
Comment (akosbalasko) [inline main.ts:151]: I think 'valueOf()' is not required, simply use this.settings.indexTagValue

File: main.ts [hunk 11]
```diff
@@ -94,99 +125,100 @@ export default class ZoottelkeeperPlugin extends Plugin {
...
+					// if one or multiple index-tags are set, they are inserted at the beginning of the index-array
+					indexContent.unshift(`---\ntags: [${tag}]\n---\n`);
+				}			
+				await this.app.vault.modify(indexTFile, indexContent.join("\n"));
+			} else {
+				throw new Error("Creation index as folder is not supported");
 			}
 		} catch (e) {
-			console.warn('Error during deletion/creation of index files');
+			console.warn("Error during deletion/creation of index files");
```
Comment (akosbalasko) [inline main.ts:160]: same here

File: main.ts [hunk 12]
```diff
@@ -94,99 +125,100 @@ export default class ZoottelkeeperPlugin extends Plugin {
...
-				throw new Error('Creation index as folder is not supported');
+			if (indexTFile instanceof TFile){
+				if (this.settings.indexTagBoolean === true) {
+					const tag = this.settings.indexTagValue.valueOf();
+					// if one or multiple index-tags are set, they are inserted at the beginning of the index-array
+					indexContent.unshift(`---\ntags: [${tag}]\n---\n`);
+				}			
+				await this.app.vault.modify(indexTFile, indexContent.join("\n"));
+			} else {
+				throw new Error("Creation index as folder is not supported");
```
Comment (akosbalasko) [inline main.ts:157]: same here

File: main.ts [hunk 13]
```diff
@@ -94,99 +125,100 @@ export default class ZoottelkeeperPlugin extends Plugin {
...
 		const parentName = this.getParentFolderName(filePath);
 
 		return `${parentPath}${this.settings.indexPrefix}${parentName}.md`;
-
-	}
+	};
 
 	getParentFolder = (filePath: string): string => {
-		const fileFolderArray = filePath.split('/');
+		const fileFolderArray = filePath.split("/");
```
Comment (akosbalasko) [inline main.ts:192]: here as well

File: main.ts [hunk 14]
```diff
@@ -94,99 +125,100 @@ export default class ZoottelkeeperPlugin extends Plugin {
...
 	isIndexFile = (file: TAbstractFile): boolean => {
 
-		return file instanceof TFile && file.name.startsWith(this.settings.indexPrefix);
-
-	}
-
+		// the original return was responsible for half the errors (and existed as a bug in OG Zoottelkeeper too)
+		// --> the problem was that the index files didn't update when the prefix was ""
+		// the if statement solves this issue by checking this case and treating it separately
+		if (this.settings.indexPrefix === "")
```
Comment (akosbalasko) [inline main.ts:211]: you can use conditional expression

File: manifest.json [hunk 15]
```diff
@@ -1,10 +1,10 @@
 {
 	"id": "zoottelkeeper-obsidian-plugin",
 	"name": "Zoottelkeeper Plugin",
-	"version": "0.6.0",
+	"version": "0.0.9",
```
Comment (akosbalasko) [inline manifest.json:4]: pls correct the version to 0.7.0

**Review by akosbalasko — CHANGES_REQUESTED** [review]

**Review by akosbalasko — APPROVED** [review]
